crates/pico_core/src/storage.rs

```diff
@@ -0,0 +1,25 @@
+use std::any::Any;
+
+use crate::{
+    container::Container,
+    database::Database,
+    epoch::Epoch,
+    key::Key,
+    node::{Dependency, DerivedNode, DerivedNodeId, SourceNode},
+    params::ParamId,
+};
+
+pub trait Storage<Db: Database + ?Sized> {
+    fn derived_nodes(&self) -> &impl Container<DerivedNodeId, DerivedNode<Db>>;
+    fn source_nodes(&self) -> &impl Container<Key, SourceNode>;
+    fn params(&self) -> &impl Container<ParamId, Box<dyn Any>>;
+    fn current_epoch(&self) -> Epoch;
+}
+
+pub trait StorageMut<Db: Database + ?Sized> {
```

Comment:
I think this should require `Storage` and have differently named methods (`derived_nodes_mut` etc)